Normalize builtin types to dtypes.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106074

Fix: #105052
Follow-up: #105588

This PR normalizes builtin Python types (e.g. `int` and `float`) into PyTorch data types
when these are passed as argument, instead of used as functions.

In summary, we:

- Implement `BuiltinVariable.as_proxy`, mapping Python types into PyTorch data types

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

ysiraichi committed:
Normalize builtin types to dtypes.

Fix: #105052
Follow-up: #105588

This PR normalizes builtin Python types (e.g. `int` and `float`) into PyTorch data types
when these are passed as argument, instead of used as functions.

In summary, we:

- Implement `BuiltinVariable.as_proxy`, mapping Python types into PyTorch data types

[ghstack-poisoned]

ysiraichi committed:
Fix lint issues. on "Normalize builtin types to dtypes."


Fix: #105052
Follow-up: #105588

This PR normalizes builtin Python types (e.g. `int` and `float`) into PyTorch data types
when these are passed as argument, instead of used as functions.

In summary, we:

- Implement `BuiltinVariable.as_proxy`, mapping Python types into PyTorch data types

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

lezcano commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -409,6 +409,15 @@ def python_type(self):
 
+    def as_proxy(self):
+        DTYPE = {
+            int: torch.int64,
```
bool and complex?

Also, is this something we already do between the Python and the C++ code in PyTorch eager? If so, it'd be good to have a look at where are we doing that, to see if there's any other logic we may need to implement.

ysiraichi replied:
Ah, forgot about `bool`. But, `complex` doesn't really work on eager. Should we enable it on dynamo?

ysiraichi replied:
I think [this](https://github.com/pytorch/pytorch/blob/cb9a4fbbf25e5f15f862d11e206a4028e9898515/torch/csrc/utils/python_arg_parser.h#L730) is the function we are looking for. Adding `bool` seems to do the job.

lezcano replied:
Thank you for digging this up!

ysiraichi committed:
Add bool support. on "Normalize builtin types to dtypes."


Fix: #105052
Follow-up: #105588

This PR normalizes builtin Python types (e.g. `int` and `float`) into PyTorch data types
when these are passed as argument, instead of used as functions.

In summary, we:

- Implement `BuiltinVariable.as_proxy`, mapping Python types into PyTorch data types

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge -i

lezcano commented:
@pytorchbot merge -i

ysiraichi commented:
@pytorchbot merge -i

ysiraichi commented:
@pytorchbot merge -i

ysiraichi committed:
Rebased. on "Normalize builtin types to dtypes."


Fix: #105052
Follow-up: #105588

This PR normalizes builtin Python types (e.g. `int` and `float`) into PyTorch data types
when these are passed as argument, instead of used as functions.

In summary, we:

- Implement `BuiltinVariable.as_proxy`, mapping Python types into PyTorch data types

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Normalize builtin types to dtypes."


Fix: #105052
Follow-up: #105588

This PR normalizes builtin Python types (e.g. `int` and `float`) into PyTorch data types
when these are passed as argument, instead of used as functions.

In summary, we:

- Implement `BuiltinVariable.as_proxy`, mapping Python types into PyTorch data types

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi commented:
@pytorchbot merge -i

OrenLeung commented:
thanks for this PR <3
